File: workers/src/utils/types/entities/game.ts
import { GameStoreMetaData } from "../../../steam/gameMetadataFetcher.js";
import { eCurrency } from "../enums/eCurrency.js";
import { isEnumValue, isNumberArray, isObject, isStringArray } from "../typeGuards.js";
import { Entity, isEntity } from "./entity.js";
import { GameOffer, isGameOffer } from "./gameOffer.js";

export interface Game extends Entity {
  name: string;
  steamUrl: string;
  steamID: number;
  inPackages: number[] | null;
  isDLC: boolean;
  description: string | null;
  headerImage: string | null;
  offers: GameOffer[] | null;
  isReleased: boolean;
  initialPrice: number | null;
  initialCurrency: eCurrency | null;
  storeMetadata?: GameStoreMetaData | null;
}

function isGameStoreMetadata(value: unknown): value is GameStoreMetaData {
  if (!isObject(value)) return false;

  return (
    isStringArray(value.tags) &&
    isStringArray(value.genres) &&
    typeof value.positiveReviews === "number" &&
    typeof value.negativeReviews === "number" &&
    typeof value.positiveReviewsPercent === "number"
  );
}

export function isGame(value: unknown): value is Game {
  if (!isObject(value) || !isEntity(value)) return false;

  return (
    typeof value.name === "string" &&
    typeof value.steamUrl === "string" &&
    typeof value.steamID === "number" &&
    (isNumberArray(value.inPackages) || value.inPackages === null) &&
    typeof value.isDLC === "boolean" &&
    (typeof value.description === "string" || value.description === null) &&
    (typeof value.headerImage === "string" || value.headerImage === null) &&
    ((Array.isArray(value.offers) && value.offers.every(isGameOffer)) || value.offers === null) &&
    typeof value.isReleased === "boolean" &&
    (typeof value.initialPrice === "number" || value.initialPrice === null) &&
    (isEnumValue(eCurrency, value.initialCurrency) || value.initialCurrency === null) &&
    (
      value.storeMetadata === undefined ||
      value.storeMetadata === null ||
      isGameStoreMetadata(value.storeMetadata)
    )
  );
}

File: src/GamesFinder.Orchestrator.Domain/Classes/Entities/Game.cs
using GamesFinder.Orchestrator.Domain.Enums;
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace GamesFinder.Orchestrator.Domain.Classes.Entities;

public class Game(
  string name,
  long steamID,
  List<GameOffer>? initialOffers = null,
  string? description = null,
  string? steamUrl = null,
  string? headerImage = null,
  Game.GameStoreMetadata? storeMetadata = null
  ) : Entity
{
  [BsonElement("name")]
  public string Name { get; set; } = name;

  [BsonElement("steam_url")]
  public string? SteamURL { get; set; } = steamUrl;
  [BsonElement("steam_id")]
  public long SteamID { get; set; } = steamID;
  [BsonElement("in_packages")]
	public List<long> InPackages { get; set; } = new();
	[BsonElement("is_DLC")]
	public bool IsDLC { get; set; }
  [BsonElement("description")]
  public string? Description { get; set; } = description;
  [BsonElement("header_image")]
  public string? HeaderImage { get; set; } = headerImage;
  [BsonIgnore]
  public List<GameOffer> Offers { get; set; } = initialOffers ?? new();
  [BsonElement("is_released")]
  public bool IsReleased { get; set; }
  [BsonElement("initial_price")]
  public decimal? InitialPrice {get; set;}
  [BsonElement("initial_currency")]
  public ECurrency? InitialCurrency {get; set;}
  [BsonElement("store_metadata")]
  public GameStoreMetadata? StoreMetadata { get; set; } = storeMetadata;

  public override string ToString()
  {
    return "🎮:\n" + base.ToString() + $"Name: {Name}, SteamID: {SteamID}, Offers count: {Offers?.Count()},\nDescription: {Description?.Substring(0, Math.Min(50, Description.Length))}..., SteamURL: {SteamURL}\nMetadata: {StoreMetadata}";
  }

  public record GameStoreMetadata
  {
    [BsonElement("tags")]
    public IEnumerable<string> Tags { get; init; } = [];
    [BsonElement("genres")]
    public IEnumerable<string> Genres { get; init; } = [];
    [BsonElement("positive_reviews")]
    public int PositiveReviews { get; init; }
    [BsonElement("negative_reviews")]
    public int NegativeReviews { get; init; }
    [BsonElement("positive_reviews_percent")]
    public double PositiveReviewsPercent { get; init; }

    public override string ToString()
    {
      return $"Tags: {string.Join(", ", Tags)}, Genres: {string.Join(", ", Genres)}, PositiveReviews: {PositiveReviews}, NegativeReviews: {NegativeReviews}, PositiveReviewsPercent: {PositiveReviewsPercent}%";
    }
  }
}
